src/stp/plays/defensive/DefendPass.cpp

```diff
@@ -53,29 +76,53 @@ void DefendPass::calculateInfoForRoles() noexcept {
+    stpInfos["defender_1"].setPositionToDefend(field.getOurTopGoalSide());
+    stpInfos["defender_1"].setBlockDistance(BlockDistance::PARTWAY);
+
+    stpInfos["defender_2"].setPositionToDefend(field.getOurBottomGoalSide());
+    stpInfos["defender_2"].setBlockDistance(BlockDistance::PARTWAY);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToBall && enemyRobot->getId() == enemyClosestToBall.value()->getId(); });
+
+    auto enemyClosestToOurGoalOne = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoalOne && enemyRobot->getId() == enemyClosestToOurGoalOne.value()->getId(); });
+
+    auto enemyClosestToOurGoalTwo = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+
+    erase_if(enemyRobots, [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoalTwo && enemyRobot->getId() == enemyClosestToOurGoalTwo.value()->getId(); });
 
-    auto enemyClosestToOurGoal = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+    stpInfos["defender_helper_1"].setPositionToDefend(enemyClosestToOurGoalOne->get()->getPos());
+    stpInfos["defender_helper_1"].setBlockDistance(BlockDistance::HALFWAY);
 
-    enemyRobots.erase(std::remove_if(enemyRobots.begin(), enemyRobots.end(),
-                                     [&](const auto enemyRobot) -> bool { return enemyClosestToOurGoal && enemyRobot->getId() == enemyClosestToOurGoal.value()->getId(); }),
-                      enemyRobots.end());
+    stpInfos["defender_helper_2"].setPositionToDefend(enemyClosestToOurGoalTwo->get()->getPos());
+    stpInfos["defender_helper_2"].setBlockDistance(BlockDistance::HALFWAY);
 
-    auto remainingEnemy = world->getWorld()->getRobotClosestToPoint(field.getOurGoalCenter(), enemyRobots);
+    std::map<double, Vector2> enemyMap;
 
-    stpInfos["defender_1"].setPositionToDefend(enemyClosestToOurGoal->get()->getPos());
-    stpInfos["defender_1"].setBlockDistance(BlockDistance::HALFWAY);
+    for (auto enemy : enemyRobots) {
+        double score = FieldComputations::getPercentageOfGoalVisibleFromPoint(field, true, enemy->getPos(), world->getWorld().value(), enemy->getId(), false) *
+                       FieldComputations::getDistanceToGoal(field, false, enemy->getPos());
+        enemyMap.insert({score, enemy->getPos()});
+    }
 
-    stpInfos["defender_2"].setPositionToDefend(remainingEnemy->get()->getPos());
-    stpInfos["defender_2"].setBlockDistance(BlockDistance::HALFWAY);
+    for (int i = 1; i <= 3; i++) {
+        stpInfos["midfielder_" + std::to_string(i)].setPositionToDefend(enemyMap.rbegin()->second);
+        stpInfos["midfielder_" + std::to_string(i)].setBlockDistance(BlockDistance::HALFWAY);
+
+        if (!enemyMap.empty()) {
+            enemyMap.erase(prev(enemyMap.end()));
+        } else {
```

Comment:
When this is true (enemyMap is empty), wouldn't the code in line 117 still try to access this empty map? Perhaps this check for emptiness should be done at the start of the loop.